pagination bug fix, ui tweaks on next/prev page

alexhoffman-dev opened the pull request:
Fixes: https://github.com/TellemBigAlsentya/local-storage-favorites/issues/5

buttons, styling tweaks

alexhoffman-dev committed:
pagination bug fix, ui tweaks on next/prev page
buttons, styling tweaks

Luke9389 commented on app.js:
```diff
@@ -88,16 +92,16 @@ async function getData(queryParam, destination) {
     collection.forEach((character) => {
         if(destination === 'grid') {
             renderCard(character);
+            disablePageButtons();
```
I don't think this is the correct place to be calling `disablePageButtons`.
Notice that this is inside of a `forEach` loop, so we are going to be calling this function for every single card.

alexhoffman-dev replied:
Totally agree -- I do think that this function should be called within `showData`, so I've moved it to the end of the function after the forEach loop concludes.

Luke9389 replied:
> so I've moved it to the end of the function after the forEach loop concludes.

I don't see that here. Have you pushed that change?

alexhoffman-dev replied:
Changes were just pushed -- I made the mistake of never updating my PR after testing in VS Code. I assume that the way to do that is either to make small changes using the VS Code editor tool or to simply stage and push said changes via VS Code?

Luke9389 commented on app.js:
```diff
@@ -161,8 +165,21 @@ function toggleHeartDisplay(event) {
     return Boolean(favorites.split(',').find(elem => elem == id));
 }
+
+function paginationCheck(pageButton) {  
+    if(pageButton === 'next-page' && currentPageData.next) {
+        return true; 
+    } else if(pageButton === 'previous-page' && currentPageData.prev) {
+        return true;
+    } else {return false;}
+}
```
```suggestion
function paginationCheck(pageButton) {  
    return pageButton === 'next-page' ? currentPageData.next : currentPageData.prev;
}
```

Luke9389 replied:
I've given a refactoring suggestion for this function above. I'll give some context here.

If you ever have code of the form 
```js
if (myCondition) {
    return true
} else {
    return false
}
```
that's the same as doing this
```js
return myCondition;
```
**_as long as `myCondition` is actually a boolean._** 

---

In my suggested version of this function, we actually end up returning the value of `currentPageData.next` and `currentPageData.prev`. If we wanted to only ever return a pure boolean we could do something like:

```js
function paginationCheck(pageButton) {
    const pageData = pageButton === 'next-page' ? currentPageData.next : currentPageData.prev;
    return Boolean(pageData)
}
```

But since we only ever call this function inside of an `if` condition, I think it's fine to return truthy values instead of booleans.

alexhoffman-dev replied:
That makes sense -- I agree that since the truthiness of `pageData` is being checked, the returned value doesn't necessitate a boolean. Very sleek one-liner to boot 🤠

Luke9389 commented on app.js:
```diff
@@ -161,8 +165,21 @@ function toggleHeartDisplay(event) {
+    } else {return false;}
+}
+
+function disablePageButtons() {
+    nextPage.disabled = currentPageData.next ? false : true; 
+    prevPage.disabled = currentPageData.prev ? false : true; 
+}
```
```suggestion
function disablePageButtons() {
    nextPage.disabled = !currentPageData.next;
    prevPage.disabled = !currentPageData.prev;
}
```

Luke9389 commented on styles.css:
```diff
@@ -271,6 +271,16 @@ hr {
+    opacity: 20%;
+    fill: gray;
+}
+button:disabled:hover {
+    cursor: not-allowed!important;
+    opacity: 20%!important;
+    fill: gray!important;
+    border: solid 3px black!important;
+}
```
Are the `!important` keywords necessary here? 

You could try switching the position of this css rule with `button:disabled` and see if you can take them out. 

If you need to leave them in, toss a space between the value and the `!important` keyword

Example
`cursor: not-allowed !important`

alexhoffman-dev replied:
I tried re-arranging these styles but since the next/previous page buttons have styling as well as styling on hover that are designated by their id's ie. `id=previous-page`, the scope overrules the `button:disabled` pseudo-class. 

I'll take a look and see if I can think of a smoother way to do that without the `!important` tag

Luke9389 replied:
Word. Sounds good. It's not a huge deal, so don't spend too much time on it. Keeping them is fine.

Luke9389 replied:
this one isn't resolved yet

Luke9389 commented on app.js:
```diff
@@ -29,12 +31,13 @@ async function load() {
-    if (currentPage < 1) { currentPage = 1; }
-    await getData(`?page=${currentPage}`, 'grid');
-    await showData('grid');
+    let targetID = event.target.id; 
+    if(paginationCheck(targetID)) {
+        currentPage = targetID === 'next-page' ? currentPage + 1 : currentPage - 1;
+        await getData(`?page=${currentPage}`, 'grid');
+        await showData('grid');
+    } 
+
```
```suggestion
    let targetID = event.target.id; 
    if (!paginationCheck(targetID)) {
        return;
    }
    
    currentPage = targetID === 'next-page' ? currentPage + 1 : currentPage - 1;
    await getData(`?page=${currentPage}`, 'grid');
    await showData('grid');
```

It's better to have an early return than to wrap the contents of a function inside of a condition. 

One way you can keep an eye out for this is that the end of the code block after the `if` is also the end of your function. 

Example of what not to do:
```js
function something(event) {
    const myCondition = event.condition
    if (myCondition) {
        // All of the contents of the something function are in here
        code
        code
        code
    } // end of code block for if statement
} // end of code block
```

Example of what to do:
```js
function something(event) {
    const myCondition = event.condition
    if (!myCondition) {
        return; // early return 
    }
    
    // rest of the function
    code
    code
    code
}
```

alexhoffman-dev replied:
Noted -- that makes sense. This a good example of how to make my code flatter which is definitely something I appreciate tips for.

Luke9389 commented on index.html:
```diff
@@ -15,14 +15,15 @@ <h1> Rick and Morty Character Index </h1>
                 <input type="search" id="filter-search" placeholder="name"></input>
-                <div id="previous-page" class="changePage">Previous - Page</div>
-                <div id="next-page" class="changePage">Next - Page</div>
+                <button id="previous-page" class="changePage">Previous - Page</button>
```
```suggestion
                <button disabled id="previous-page" class="changePage">Previous - Page</button>
```

You can do this instead of disabling it at the beginning of your JavaScript file.

Luke9389 reviewed (changes requested):
Good start! Here are some suggestions. 

Btw every time i use the `Commit Suggestion` button in GH something breaks. I don't recommend it 😆

alexhoffman-dev committed:
Update app.js

Co-authored-by: Luke Donahue <49007721+Luke9389@users.noreply.github.com>

alexhoffman-dev committed:
Update app.js

Co-authored-by: Luke Donahue <49007721+Luke9389@users.noreply.github.com>

alexhoffman-dev committed:
Update app.js

Flatter refactor

Co-authored-by: Luke Donahue <49007721+Luke9389@users.noreply.github.com>

alexhoffman-dev commented on app.js:
```diff
@@ -2,23 +2,25 @@ const RickAndMortyAPI = 'https://rickandmortyapi.com/api/character/';
-
+// prev page starts disabled to prevent decrementing page below 1 
+prevPage.disabled = true; 
 load();
```
remove this line -- added `disabled` pseudo-class to HTML element

alexhoffman-dev committed:
Update index.html

Co-authored-by: Luke Donahue <49007721+Luke9389@users.noreply.github.com>

alexhoffman-dev committed:
Merge branch 'main' into Alex-issue-resolutions

Luke9389 commented:
OK, so what's left here to finish this one off?

alexhoffman-dev committed:
Update styles.css

Updated code -- I made these changes on my working branch in VS Code to do some testing and never updated the PR.

alexhoffman-dev committed:
Update app.js

A similar story here -- I never updated the code on my this PR. 

Code changes: 
- added new architecture for page function
- moved the disablePage function to the end of showData 
- added bug fix to favorited card DOM elements so that on re-load, they persist with the 'favorited' class --> toggleHeartDisplay functions properly now.

alexhoffman-dev commented on app.js:
```diff
@@ -2,10 +2,11 @@ const RickAndMortyAPI = 'https://rickandmortyapi.com/api/character/';
 // ARM BUTTONS
 const searchInput = document.getElementById('filter-search');
-const favoriteBar = document.getElementById('favorite-container');
+const favoriteBar = document.getElementById('favorite-card-container');
```
Created a new container for the favorite cards so that the 'Favorites' header isn't cleared once a favorite card is loaded

Luke9389 commented:
Nice! Looks good! Before you merge it, be sure you've tested it all (heh). Don't wanna push any bugs to the live site :-)

TellemBigAlsentya commented:
@Luke9389 I'm going to need to stick to this one for the time being since the Netlify deploy is not happy about an error that my live port via Chrome wasn't concerned about.
